sw/device/lib/dif/dif_alert_handler.h

```diff
@@ -0,0 +1,649 @@
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Generic enable/disable enumeration.
+ */
+typedef enum dif_alert_handler_enable {
+  kDifAlertHandlerEnabled,
+  kDifAlertHandlerDisabled,
+} dif_alert_handler_enable_t;
+
+/**
+ * An alert class.
+ *
+ * An alert class roughly specifies how to deal with an alert. The class
+ * determines which interrupt handler is fired for an alert, as well as the
+ * fine-grained details of the escalation policy, for when the processor
+ * fails to respond to an alert quickly enough.
+ *
+ * Alert classes serve as the alert handler's IRQ types. There is one IRQ for
+ * each class.
+ */
+typedef enum dif_alert_handler_class {
+  kDifAlertHandlerClassA,
+  kDifAlertHandlerClassB,
+  kDifAlertHandlerClassC,
+  kDifAlertHandlerClassD,
+} dif_alert_handler_class_t;
+
+/**
+ * An alert, identified by a numeric id.
+ *
+ * Alerts are hardware-level events indicating that something catastrophic
+ * has happened. The alert handler consumes alerts, classifies them to a
+ * particular `dif_alert_handler_class_t`, and uses policy information attached
+ * to that class to handle it.
+ *
+ * The number of alerts is configurable at hardware-synthesis time, and is
+ * specified by the software when initializing a `dif_alert_handler_t`.
+ */
+typedef uint32_t dif_alert_handler_alert_t;
+
+/**
+ * A local alert originating from within the alert handler itself.
+ *
+ * A local alert is exactly the same as a normal `dif_alert_handler_alert_t`,
+ * except that they use different functions for setting up classification and
+ * for getting causes.
+ */
+typedef enum dif_alert_handler_local_alert {
+  kDifAlertHandlerLocalAlertAlertPingFail,
+  kDifAlertHandlerLocalAlertEscalationPingFail,
+  kDifAlertHandlerLocalAlertAlertIntegrityFail,
+  kDifAlertHandlerLocalAlertEscalationIntegrityFail,
+} dif_alert_handler_local_alert_t;
+
+/**
+ * An escalation signal, identified by a numeric id.
+ *
+ * An escalation signal is a generic "response" to failing to handle a
+ * catastrophic event. What each signal means is determined by the chip.
+ *
+ * A `dif_alert_handler_class_t` can be configured to raise escalation signals
+ * as part of its policy.
+ */
+typedef uint32_t dif_alert_handler_signal_t;
+
+/**
+ * An alert class state
+ *
+ * Every class begins in an idle state. When sufficient alerts have
+ * accumulated, the following procedure occurs:
+ * - The alert class IRQ is fired to the processor.
+ * - If the processor fails to clear the IRQ within a set timeout,
+ *   escalation phase 0 begins, potentially triggering an escalation signal.
+ * - After some time, phase 1 begins, and so on.
+ * - Once phase 3's duration ends, the class enters the "terminal" state, which
+ *   cannot be exited except by reset.
+ *
+ * At any point, software may end the escalation protocol by calling
+ * `dif_alert_handler_phase_escalation_clear()` (unless clearing is disabled).
+ * Succeessfully calling this function, or clearing the IRQ on time, will reset
```

Comment:
Maybe it is worth mentioning that this functions is not guaranteed to clear the escalation due to timings (if the escalation process finishes after the function have been called)?

Reply:
I think that's kind of an unimportant edge case, but we can include the comment.